[라테] 로또 미션 제출합니다.

minuyim opened the pull request:
로또 미션 제출드립니다.

감사합니다!

include42 committed:
[docs] README에 요구사항 및 구현단위 작성

README 요구사항을 간략히 정리하고 테스트를 위한 구현 단위 일부 작성

include42 committed:
[feat] Money 클래스 구현 및 MoneyTest 구현

Money 로또 구입 로직에서 금액을 담당
MoneyTest Money에 대한 테스트(생성, 입력, 예외처리) 구현

include42 committed:
[refactor] Money 클래스 refactor

Money 예외처리 메소드 분리 및 상수 처리

include42 committed:
[docs] README 구현단위 작성

README 로또번호 클래스 관련 테스트단위 작성

include42 committed:
[feat] LottoNumber 클래스 구현

LottoNumber 생성자 및 equals 로직 구현
LottoNumberTest 생성, 예외처리, 비교연산 테스트 수행

include42 committed:
[feat] LottoNumbers 클래스 구현

LottoNumbers 생성자 및 공통 요소 검사 기능
LottoNumbersTest 생성, 예외처리, 공통요소 count 테스트 수행

include42 committed:
[refactor] LottoNumbers 상수 처리

include42 committed:
[feat] NumberGenerator 인터페이스 구현 및 RandomNumberGenerator 구현

NumberGenerator LottoNumbers 생성을 위한 인터페이스
RandomNumberGenerator LottoNumbers를 랜덤으로 생성하기 위한 클래스

include42 committed:
[feat] LottoNumbersFactory 클래스 구현

LottoNumbersFactory create 메소드를 NumberGenerator를 받아 LottoNumbers 생성
LottoNumbersFactoryTest 생성테스트 구현

include42 committed:
[feat] LottoRank enum 구현

LottoRank 게임 결과값을 이용해 Rank 생성
LottoRankTest Rank 생성 테스트

include42 committed:
[feat] LottoRank enum 구현

LottoRank 게임 결과값을 이용해 Rank 생성

include42 committed:
[feat] LottoWinner 클래스 구현 및 코드 리팩토링

LottoWinner 당첨번호와 다른 Numbers를 비교하는 클래스 구현
LottoWinnerTest 생성 및 랭크 반환 테스트 수행
ALL 컨벤션 준수를 위한 코드 리팩토링

include42 committed:
[docs] README 구현단위 작성

README 열거형 및 당첨번호 관련하여 구현내용 작성

include42 committed:
[feat] LottoResult 클래스 구현 및 코드 리팩토링

LottoResult 당첨 결과를 저장하는 클래스 구현
LottoResultTest LottoResult의 생성, 입출력, 예외 테스트 검증
ResultCount LottoResult 내의 당첨 횟수 기록을 담당하는 클래스
ResultCountTest ResultCount의 생성, 증가 메서드 검증

include42 committed:
[feat] 뷰 클래스 작성

InputView 입력을 담당하는 뷰 클래스 구현
OutputView 출력을 담당하는 뷰 클래스 구현

include42 committed:
[feat] 컨트롤러 클래스 작성 및 작동 확인

LottoController 컨트롤러 역할을 하는 클래스 구현, 작동 확인
Application 메인 로직을 수행하는 클래스
ALL MVC 구조에 맞춰 코드 일부 리팩토링

include42 committed:
[refactor] LottoNumber 클래스 리팩토링

LottoNumber 캐시값을 통해 객체 재활용
ALL LottoNumber 구조변경에 따른 리팩토링

include42 committed:
[refactor] 패키지 리팩토링

ALL 패키지 리팩토링

include42 committed:
[refactor] 코드 리팩토링

ALL 컨벤션 준수를 위한 코드 리팩토링

include42 committed:
[refactor] 정적 클래스의 생성자 호출 방지 구현

LottoNumbersFactory,Input,Output 생성자 호출 시 예외 발생

include42 committed:
[refactor] LottoController 내 메서드 분리

LottoController 기능에 따라 메서드 분리

include42 committed:
[refactor] 테스트코드 내 코드 리팩토링

TEST 테스트코드 내의 컨벤션 미준수에 따른 코드 리팩토링

minuyim committed:
[refactor] List<Integer>의 경우 Factory에서 직접 input을 받도록 수정

LottoNumbersFactory create overloading 기능 추

minuyim committed:
[refactor] 기타 컨벤션 준수

ALL 컨벤션 준수를 위한 refactor

minuyim committed:
[refactor] LottoGame 클래스 추가 및 LottoResult 구조 수정

LottoGame List<LottoNumbers>의 일급컬렉션 생성
LottoResult Map구조 수정 및 메소드 구조 수정

minuyim committed:
[refactor] 컨벤션 준수 및 클래스 구조 수정

minuyim committed:
[fix] LottoResult에서 당첨되지 않은 Rank를 출력하지 않는 문제를 수정

minuyim committed:
[docs] LottoResult 변경에 따른 README 수정

minuyim committed:
[refactor] 메소드 순서 수정 및 LottoGame null가드 추가

minuyim committed:
[refactor] LottoNumbers List를 Set으로 수정

young891221 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,36 @@
+        OutputView.printResultTitle();
+        OutputView.printLottoResult(lottoResult);
+        OutputView.printEarning(money.calculateEarningRate(lottoResult.calculateEarning()));
+    }
```
`OutputView.printAll()`과 같이 처리해 보는건 어떨까요?

minuyim replied:
마지막 Result에 대한 내용을 출력하는 부분이니 한 곳에 묶는게 보기 좋을 것 같습니다.

young891221 commented on src/main/java/domain/lotto/LottoGame.java:
```diff
@@ -0,0 +1,46 @@
+
+public class LottoGame {
+    private static final int COUNT_NUMBER = 1;
+    private static final String ERROR_NULL_MESSAGE = "1부터 45 사이의 숫자만 입력 가능합니다.";
```
`ERROR_NULL_MESSAGE` 네이밍와 value값이 일치하지 않습니다.

minuyim replied:
잘못된 메세지 사용으로 혼동을 드린 점 죄송합니다...

young891221 commented on src/main/java/domain/lotto/LottoGame.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    private void validateNullValue(List<LottoNumbers> lottoGame) {
+        if (Objects.isNull(lottoGame)) {
```
메시지처럼 1~45까지의 수를 체크하는게 맞아 보입니다.
아니면 null에 대한 체크메시지를 반환하도록 해야 합니다.

young891221 replied:
LottoNumber에서 숫자 범위를 이미 체크를 하고 있군요.  여기서는 무의미해 보이네요.

minuyim replied:
잘못된 메세지 사용으로 혼동을 드린 점 죄송합니다...(2)

young891221 commented on src/main/java/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,44 @@
+    private static final String ERROR_BOUND_MESSAGE = "1부터 45 사이의 숫자만 입력 가능합니다.";
+    private static final int LOTTO_UNDER_BOUND = 1;
+    private static final int LOTTO_UPPER_BOUND = 45;
+    private static final Map<Integer, LottoNumber> cache = new HashMap<>();
```
LottoNumber를 자체적으로 캐시하는 신박한(?) 방법을 사용하셨군요.
LottoNumber는 자신의 고유값 즉 int만 할당시키고 LottoNumberFactory와 같은 클래스로 LottoNumber를 생성해 주도록 처리하면 어떨까요?ㅎㅎ

minuyim replied:
캐시생성 및 보유에 대한 피드백 반영하여 수정하였습니다.

young891221 commented on src/main/java/domain/lotto/generator/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,30 @@
+    private static final int END_INDEX = 6;
+
+    static {
+        numbers = IntStream.range(START_LOTTO_INCLUSIVE, END_LOTTO_EXCLUSIVE)
```
rangeClosed()를 사용해서 END_LOTTO_EXCLUSIVE를 45로 변경하는 것이 일관성에 좋아 보입니다.
(rangeClosed가 맞는지 몰르겠네요. 기억나는데로ㅎㅎ;;)

minuyim replied:
INCLUSIVE, EXCLUSIVE는 고정되는 줄 알았는데 rangeClosed라는 기능을 통해서도 구현 가능하군요. stream에는 다양한 기능이 있는 것을 알게 되었습니다. 감사합니다.

young891221 commented on src/main/java/domain/lotto/generator/NumberGenerator.java:
```diff
@@ -0,0 +1,9 @@
+import java.util.SortedSet;
+
+public interface NumberGenerator {
+    SortedSet<LottoNumber> create();
```
오 SortedSet 사용은 굿입니다. 👍

minuyim replied:
페어의 도움으로 바꿀 수 있었습니다. 감사합니다!

young891221 reviewed (changes requested):
안녕하세요, 라테!
전체적으로 구조나 테스트 작성 방식이 보기 좋습니다. 👍 
몇 가지 개선사항만 처리하면 훨씬 좋아보이네요. 요것들만 반영해 주시면 머지 하겠습니다:)

minuyim committed:
[refactor] 피드백 반영 및 for문을 stream으로 대체해보기

minuyim committed:
[refactor] LottoNumberFactory 클래스 추가 및 LottoNumber 구조 수정

LottoNumberFactory 로또번호의 캐시값을 저장하고 로또번호를 반환하는 클래스
LottoNumber 캐시값을 넘겨주고 wrapper class 책임만 수행하도록 조정

minuyim committed:
[docs] README update

현재 구현 상황과 맞지 않는 내용 수정

minuyim committed:
[refactor] orElse를 orElseGet으로 바꾸어보기

orElse에 메서드가 있을 경우 일단 실행하고 Optional 객체가 비어있으면 반환된다
따라서 orElseGet이 좀 더 안전할 수 있다

minuyim committed:
[refactor] 기타 수정사항 반영

메소드 체이닝 줄여보기
가독성 좋게 메소드 수정하기

minuyim committed:
[refactor] LottoGame 인자 private 추가

minuyim committed:
[refactor] formatting 적용 및 생략된 테스트 추가

LottoRank multiplyCount 테스트 추가

young891221 reviewed (approved):
피드백 반영하시느라 고생하셨습니다. 라테! 💯 
훨씬 보기 좋아진 것 같습니다.
머지할게요~!
